tests/unit/test_gce_clients.py

```diff
@@ -0,0 +1,344 @@
+                            'subnetwork': 'subnetwork/url/string',
+                            'networkIP': '192.168.0.1',
+                            'name': 'test-network',
+                            'accessConfigs': [{
+                                'type': 'ONE_TO_ONE_NAT',
+                                'name': 'EXTERNAL NAT',
+                                'natIP': '1.1.1.1',
+                                'kind': 'compute#accessConfig'
+                            }],
+                        }],
+                        'metadata': {
+                            'items': [
+                                {'key': 'default', 'value': 'true'}
+                            ],
+                        },
+                    }],
+                    'warning': {
+                        'warning': 'object'
+                    }
+                }
+            },
+            'nextPageToken':
+            None
+        }
+
+    @pytest.mark.asyncio
+    async def test_list_instances_from_one_page(
+            self, compute_rsp, compute_base_url, get_gce_client):
+        gcp_client = get_gce_client(gce_clients.GCPClient)
+        with aioresponses() as m:
+            m.get(
+                compute_base_url + 'maxResults=10&filter=zone eq us-west9-z',
+                payload=compute_rsp)
+
+            results = await gcp_client.list_instances(
+                'test-project',
+                page_size=10,
+                instance_filter='zone eq us-west9-z')
+
+            expected_results = [{
+                'hostname': 'instance-1',
+                'internal_ip': '192.168.0.1',
+                'external_ip': '1.1.1.1'
+            }]
+        assert results == expected_results
+
+    @pytest.mark.asyncio
+    async def test_list_instances_filter_by_tag(
+            self, compute_rsp, compute_base_url, get_gce_client, caplog):
+        gcp_client = get_gce_client(gce_clients.GCPClient)
+        caplog.set_level(logging.ERROR)
+        blacklisted_instance = \
+            copy.deepcopy(compute_rsp['items']['us-west1-z']['instances'][0])
+        blacklisted_instance['name'] = 'instance-2'
+        blacklisted_instance['tags']['items'] = ['bl-tag']
+        compute_rsp['items']['us-west1-z']['instances'].append(
+            blacklisted_instance)
+
+        gcp_client.blacklisted_tags.append('bl-tag')
+
+        with aioresponses() as m:
+            m.get(
+                compute_base_url + 'maxResults=10&filter=',
+                payload=compute_rsp)
+
+            results = await gcp_client.list_instances(
+                'test-project', page_size=10)
+
+            expected_results = [{
+                'hostname': 'instance-1',
+                'internal_ip': '192.168.0.1',
+                'external_ip': '1.1.1.1'
+            }]
+        assert results == expected_results
+        assert len(caplog.records) == 1
+
+    @pytest.mark.asyncio
+    async def test_list_instances_filter_by_metadata(
```

Comment:
Can this test + `test_list_instances_filter_by_tag` + `test_list_instances_from_one_page` be parametrized at all? Seems like a lot of repeated code.

Reply:
:+1: Packed all of them down into one test!